### PR TITLE
Add Line Allgather Support

### PR DESCRIPTION
**Pull request by SeanNijjar**
Line all-gather support is required to enable model bringup on t7000 galaxy systems. 

This part's mainly relevant to @tt-aho 
This change looks pretty invasive to `all_gather_op_multi_core.cpp` but it's mostly indentation with select changes made to enable linear. At a high-level, I reused most of the ring-all-gather host code and "tricked" it into disabling bidirectional so it would send the full tensor in the clockwise direction. I then run the host code again (in a direction loop), with bidirectional disabled, but with ring direction set to counter-clockwise to force the full tensor to also be sent in the counter-clockwise direction.

Future work to add new ops (reduce-scatter), cut down this large function into constituent parts and more reusable components (across ops), although you can start to see the beginnings of that in this PR (like in the addition of the EDM builder component). This should hopefully cut down the size of future PRs.

### CODE REVIEW

**Comment by SeanNijjar** [comment]
post-commit: https://github.com/tenstorrent-metal/tt-metal/actions/runs/8574013728
multi-nebula unit test: https://github.com/tenstorrent-metal/tt-metal/actions/runs/8574038300

**Review by eyonland — APPROVED** [review]

**Comment by SeanNijjar** [comment]
Hey @TT-BrianLiu, @tt-aho, @mywoodstock, github requires review from you before this can be merged. Can you please take a look when you get a chance? Thanks.

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py:404]: Could you use loguru instead of prints here?

**Review by tt-aho — APPROVED** [review]

**Pull request merged**
